File: zul/src/archive/web/js/zul/html/tab.js
/* tab.js

{{IS_NOTE
	$Id: tab.js,v 1.13 2006/05/10 10:04:02 tomyeh Exp $
	Purpose:
		
	Description:
		
	History:
		Tue Jul 12 17:12:50     2005, Created by tomyeh@potix.com
}}IS_NOTE

Copyright (C) 2005 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 2.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
////
// tab //
function zkTab() {};

/** Selects the specified tab. */
zkTab.onclick = function (evt) {
	if (!evt) evt = window.event;
	var tab = zkau.getParentByType(Event.element(evt), "Tab");
	zkTab.selTab(tab);
};

/** Returns the selected tab by giving any tab as the reference point. */
zkTab._getSelTab = function (tab) {
	var tabboxId = tab.getAttribute("zk_box");
	if (tabboxId) {
		var tabbox = $(tabboxId);
		if (tabbox.getAttribute("zk_accd") == "true")
			return zkTab._getSelTabFromTop(tabbox, tabboxId);
	}

	for (var node = tab; (node = node.nextSibling) != null;)
		if (node.getAttribute && node.getAttribute("zk_sel") == "true")
			return node;
	for (var node = tab; (node = node.previousSibling) != null;)
		if (node.getAttribute && node.getAttribute("zk_sel") == "true")
			return node;
};
/** Returns the selected tab by specified any HTML tag containing it. */
zkTab._getSelTabFromTop = function (node, tabboxId) {
	if (zk.getCompType(node) == "Tab"
	&& node.getAttribute("zk_box") == tabboxId)
		return node.getAttribute("zk_sel") == "true" ? node: null;

	for (var node = node.firstChild; node != null; node = node.nextSibling) {
		var n = zkTab._getSelTabFromTop(node, tabboxId);
		if (n) return n;
	}
	return null;
};

/** Selects the specified tab (and unselect the current tab). */
zkTab.selTab = function (tab) {
	if (!tab) return;
	tab = $(tab);

	var old = zkTab._getSelTab(tab);
	if (old != tab) {
		if (old) zkTab._setTabSel(old, false);
		zkTab._setTabSel(tab, true);
	}
};

/** Selects o unselect the specified tab. */
zkTab._setTabSel = function (tab, toSel) {
	if ((tab.getAttribute("zk_sel") == "true") == toSel)
		return; //nothing changed

	tab.setAttribute("zk_sel", toSel ? "true": "false");
	var tabreal = $(tab.id + "!real");
	if (tabreal) {
		if (toSel) {
			tabreal.className = tabreal.className + "sel";
		} else {
			var len = tabreal.className.length;
			if (len > 3)
				tabreal.className = tabreal.className.substring(0, len - 3);
		}
	}

	zkTab._changeBkgnd(tab, toSel);

	var tabbox = tab.getAttribute("zk_box");
	if (tabbox) tabbox = $(tabbox);
	var accd = tabbox && tabbox.getAttribute("zk_accd") == "true";
	var panel = $(tab.getAttribute("zk_panel"));
	if (panel)
		if (accd) action.slideDown($(panel.id + "!real"), toSel);
		else action.show(panel, toSel);

	if (!accd) {
		var tabs = zk.parentNode(zk.parentNode(tab, "TABLE"), "THEAD");
		if (tabs)
			zkTab.fixWidth(tabs.id + "!last");
	}

	if (toSel && tabbox)
		zkau.send({uuid: tabbox.id, cmd: "onSelect", data: [tab.id]},
				zkau.asapTimeout(tabbox, "onSelect"));
};
/** Changes the images in the background. */
zkTab._changeBkgnd = function (node, toSel) {
	//FUTURE: we mighit use style.backgroundImage = "url(...)" >> more standard
	if (node.background && node.background.indexOf(toSel ? "-uns": "-sel") >= 0) { //IE
		node.background = zk.renType(node.background, toSel ? "sel": "uns");
	} else if (node.getAttribute) { //Mozilla
		var img = node.getAttribute("background");
		if (img && img.indexOf(toSel ? "-uns": "-sel") >= 0)
			node.setAttribute("background", zk.renType(img, toSel ? "sel": "uns"));
	}

	for (node = node.firstChild; node; node = node.nextSibling)
		zkTab._changeBkgnd(node, toSel);
};

/** Fix the width of the last column in tabs. */
zkTab.fixWidth = function (uuid) {
	var n = $(uuid);
	if (!n) return;

	var tbl = zk.parentNode(zk.parentNode(n, "TABLE"), "TABLE");
	if ("TD" == zk.tagName(n)) { //horizontal
		var tabs = zk.parentNode(tbl, "THEAD");
		if (tabs) {
			var v = tabs.offsetWidth - tbl.offsetWidth + n.offsetWidth;
			if (v < 0) v = 0;
			n.style.width = v + "px";
		}
	} else { //vertical
		var tabs = zk.parentNode(tbl, "TD");
		if (tabs) {
			var v = tabs.offsetHeight - tbl.offsetHeight + n.offsetHeight;
			if (v < 0) v = 0;
			n.style.height = v + "px";
		}
	}
};

zkTab.init = function (cmp) {
	Event.observe(cmp, "click", function (evt) {zkTab.onclick(evt);});

	var anc = $(cmp.id + "!a");
	if (anc) {
		Event.observe(anc, "focus", function () {zkau.onfocus(anc);});
		Event.observe(anc, "blur", function () {zkau.onblur(anc);});
	}
};

////
// tabs //
function zkTabs() {}

zkTabs.init = function (cmp) {
	setTimeout("zkTab.fixWidth('"+cmp.id+"!last')", 30);
};
zkTabs.onVisi = function (cmp) {
	zkTabs.init(cmp);
};

File: zul/src/org/zkoss/zul/api/Tab.java
/* Tab.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Tue Oct 22 14:45:31     2008, Created by Flyworld
}}IS_NOTE

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 3.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
 */
package org.zkoss.zul.api;

/**
 * A tab.
 * <p>
 * Default {@link #getZclass}: z-tab. (since 3.5.0)
 * 
 * @author tomyeh
 * @since 3.5.2
 * 
 */
public interface Tab extends org.zkoss.zul.impl.api.LabelImageElement {

	/**
	 * Returns whether this tab is closable. If closable, a button is displayed
	 * and the onClose event is sent if an user clicks the button.
	 * <p>
	 * Default: false.
	 */
	public boolean isClosable();

	/**
	 * Sets whether this tab is closable. If closable, a button is displayed and
	 * the onClose event is sent if an user clicks the button.
	 * <p>
	 * Default: false.
	 * <p>
	 * You can intercept the default behavior by either overriding
	 * {@link org.zkoss.zul.Tab#onClose}, or listening the onClose event.
	 */
	public void setClosable(boolean closable);

	/**
	 * Returns the tabbox owns this component.
	 */
	public org.zkoss.zul.api.Tabbox getTabboxApi();

	/**
	 * Returns the panel associated with this tab.
	 */
	public org.zkoss.zul.api.Tabpanel getLinkedPanelApi();

	/**
	 * Sets whether this tab is selected.
	 */
	public void setSelected(boolean selected);

	/**
	 * Sets whether this tab is disabled. If a tab is disabled, then it cann't
	 * be selected or closed by user, but it still can be controlled by server
	 * side program.
	 * 
	 */
	public void setDisabled(boolean disabled);

	/**
	 * Returns the index of this panel, or -1 if it doesn't belong to any tabs.
	 */
	public int getIndex();

}
